File: frontend/lib/mosaicLayout.ts
import { MosaicTile } from "@/components/FarmMosaic"

// V2.4 — shared Farm-Mosaic layout (PROJECT_SPECIFICATION.md §V2.4–§V2.6).
// Both `FarmMosaic` (the rendering engine) and `OverlayLayer` (the tree overlay)
// must place geometry in the *same* farm-pixel coordinate space, otherwise tree
// boxes drift from their tiles. This single pure function is the one source of
// the transform; it is deliberately framework-free so it can be reused by the
// overlay without duplicating layout math or touching the rendering engine.
//
// Tile images are drawn 1:1 at their persisted `image_width/height`; column/row
// offsets are cumulative widths so mixed-size tiles stay grid-aligned (no
// overlap). A persisted tree detection's `local_pixel_*` / `bbox_*` are in that
// same tile-pixel space, so `farm = tile.(x,y) + local` aligns exactly.
export type PlacedTile = MosaicTile & {
  row: number
  col: number
  x: number
  y: number
  w: number
  h: number
}

// Defensive zero-guard for a tile's pixel dimensions only — never a layout
// fallback. Version 2 persists image_width/height (§V2.5), so a genuine mission
// always supplies them; this merely stops a degenerate 0 from collapsing the
// canvas. The layout is never recomputed or synthesized.
const MIN_TILE = 1

export function computeMosaicLayout(
  tiles: MosaicTile[],
  gap = 2
): PlacedTile[] {
  if (tiles.length === 0) return []

  const colW = new Map<number, number>()
  const rowH = new Map<number, number>()
  for (const t of tiles) {
    const row = t.grid_row as number
    const col = t.grid_col as number
    const w = Math.max(t.image_width ?? 0, MIN_TILE)
    const h = Math.max(t.image_height ?? 0, MIN_TILE)
    colW.set(col, Math.max(colW.get(col) ?? 0, w))
    rowH.set(row, Math.max(rowH.get(row) ?? 0, h))
  }

  const maxCol = Math.max(...tiles.map((t) => t.grid_col as number))
  const maxRow = Math.max(...tiles.map((t) => t.grid_row as number))

  const colX = new Array<number>(maxCol + 1)
  let x = gap
  for (let c = 0; c <= maxCol; c++) {
    colX[c] = x
    x += (colW.get(c) ?? MIN_TILE) + gap
  }
  const rowY = new Array<number>(maxRow + 1)
  let y = gap
  for (let r = 0; r <= maxRow; r++) {
    rowY[r] = y
    y += (rowH.get(r) ?? MIN_TILE) + gap
  }

  return tiles.map((t) => ({
    ...t,
    row: t.grid_row as number,
    col: t.grid_col as number,
    x: colX[t.grid_col as number],
    y: rowY[t.grid_row as number],
    w: Math.max(t.image_width ?? 0, MIN_TILE),
    h: Math.max(t.image_height ?? 0, MIN_TILE),
  }))
}

File: backend/navigation/mosaic_layout.py
"""Farm-pixel mosaic layout — backend port of the frontend single source of truth.

PROJECT_SPECIFICATION.md (§V2.4–§V2.6) and Decision 6 establish that **one** pure
function, ``computeMosaicLayout`` (``frontend/lib/mosaicLayout.ts``), owns the
transform from survey tiles to farm-pixel coordinates. Both ``FarmMosaic`` and
``OverlayLayer`` use it so tree boxes never drift from their tiles.

The Version 3 robot navigates in that **same** farm-pixel space (no SLAM, no second
coordinate system — PROJECT_SPECIFICATION.md Appendix A §A.2). To keep a single
source of truth the backend must place tiles with the *identical* algorithm the
frontend uses. This module is a faithful Python port of ``computeMosaicLayout``:
same column/row max-width model, same occupied-bounding-box fit, same ``gap``
handling. It is pure and deterministic — given the same tiles it always yields the
same placement.

Only ``SurveyTile`` geometry is consumed here (``grid_col``, ``grid_row``,
``image_width``, ``image_height``), matching the frontend exactly.
"""

from dataclasses import dataclass
from typing import Dict, List, Optional, Sequence

MIN_TILE = 1
DEFAULT_GAP = 2


@dataclass(frozen=True)
class TileGeometry:
    """Minimal tile geometry needed to place a tile in farm-pixel space."""

    id: int
    grid_col: int
    grid_row: int
    image_width: int
    image_height: int


@dataclass(frozen=True)
class PlacedTile:
    """A tile placed at its farm-pixel top-left ``(x, y)``."""

    id: int
    grid_col: int
    grid_row: int
    x: float
    y: float
    w: float
    h: float


def compute_mosaic_layout(
    tiles: Sequence[TileGeometry], gap: int = DEFAULT_GAP
) -> List[PlacedTile]:
    """Faithful backend port of ``frontend/lib/mosaicLayout.ts::computeMosaicLayout``.

    Places each tile at the cumulative column/row offset of the *occupied* farm
    rectangle (min/max grid col/row), using the per-column max width and per-row
    max height. Deterministic: identical input → identical output.
    """
    if not tiles:
        return []

    col_w: Dict[int, int] = {}
    row_h: Dict[int, int] = {}
    for t in tiles:
        w = max(t.image_width or 0, MIN_TILE)
        h = max(t.image_height or 0, MIN_TILE)
        col_w[t.grid_col] = max(col_w.get(t.grid_col, 0), w)
        row_h[t.grid_row] = max(row_h.get(t.grid_row, 0), h)

    cols = [t.grid_col for t in tiles]
    rows = [t.grid_row for t in tiles]
    min_col, max_col = min(cols), max(cols)
    min_row, max_row = min(rows), max(rows)

    col_x = [0.0] * (max_col + 1)
    x = float(gap)
    for c in range(min_col, max_col + 1):
        col_x[c] = x
        x += col_w.get(c, MIN_TILE) + gap

    row_y = [0.0] * (max_row + 1)
    y = float(gap)
    for r in range(min_row, max_row + 1):
        row_y[r] = y
        y += row_h.get(r, MIN_TILE) + gap

    placed: List[PlacedTile] = []
    for t in tiles:
        w = max(t.image_width or 0, MIN_TILE)
        h = max(t.image_height or 0, MIN_TILE)
        placed.append(
            PlacedTile(
                id=t.id,
                grid_col=t.grid_col,
                grid_row=t.grid_row,
                x=col_x[t.grid_col],
                y=row_y[t.grid_row],
                w=w,
                h=h,
            )
        )
    return placed


def tile_placement_map(
    tiles: Sequence[TileGeometry], gap: int = DEFAULT_GAP
) -> Dict[int, PlacedTile]:
    """``{tile_id: PlacedTile}`` for O(1) lookup during target resolution."""
    return {p.id: p for p in compute_mosaic_layout(tiles, gap)}


def tree_target_pixel(
    placement: Dict[int, PlacedTile],
    survey_tile_id: int,
    local_pixel_x: float,
    local_pixel_y: float,
) -> Optional[tuple]:
    """Farm-pixel target of a tree's representative observation.

    Mirrors ``OverlayLayer``: the tree centroid is the tile's placed top-left plus
    the observation's ``local_pixel_*`` offset within the tile. Returns ``None``
    when the referenced tile is not present in the placement set.
    """
    tile = placement.get(survey_tile_id)
    if tile is None:
        return None
    return (tile.x + local_pixel_x, tile.y + local_pixel_y)
